### PR TITLE
The big cooccurrence refactor

### PR DESCRIPTION
**Pull request by cjweir**
The big refactoring is basically complete.   The only draw back is that your kernel functions (if you are doing multiple window sizes) need to all be the same.  This is a numba list of functions issue.  It will take some work to add this back in I think... it can be another PR.

### CODE REVIEW

**Review by lmcinnes — APPROVED** [review]

File: vectorizers/preprocessing.py [hunk 1]
```diff
@@ -36,6 +37,33 @@ def construct_document_frequency(token_by_doc_sequence, token_dictionary):
     return doc_freq / len(token_by_doc_sequence)
 
 
+def construct_timed_document_frequency(token_by_doc_sequence, token_dictionary):
+    """Returns the frequency of documents that each token appears in.
+
+    Parameters
+    ----------
+    token_by_doc_sequence: Iterable
+        A sequence of sequences of tokens
```
Comment (jc-healy) [inline vectorizers/preprocessing.py:46]: I think this should be "a sequence of sequences of (token, timestamp) pairs." Or something like that. Correct?

File: vectorizers/preprocessing.py [hunk 2]
```diff
@@ -666,3 +692,373 @@ def preprocess_token_sequences(
...
+    """Perform a standard set of preprocessing for token sequences. This includes
+    constructing a token dictionary and token frequencies, pruning the dictionary
+    according to frequency and ignored token constraints, and editing the token
+    sequences to only include tokens in the pruned dictionary. Note that either
+    min_occurrences or min_frequency can be provided (respectively
+    max_occurences or max_frequency). If both are provided they must agree.
+
+    Parameters
+    ----------
+    token_sequences: Iterable of (tuple | list | numpy.array)
```
Comment (jc-healy) [inline vectorizers/preprocessing.py:901]: Should this be a "a list of multitoken sequences" and a few more details about the data type?

File: vectorizers/preprocessing.py [hunk 3]
```diff
@@ -666,3 +692,373 @@ def preprocess_token_sequences(
...
```
Comment (jc-healy) [inline vectorizers/preprocessing.py:722]: I think this should include a description of your timed token structure.

File: vectorizers/timed_token_cooccurrence_vectorizer.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+                target_word,
+                windows,
+                kernels,
+            )
+
+    return posterior_data
+
+
+class TimedTokenCooccurrenceVectorizer(BaseCooccurrenceVectorizer):
+    """Given a sequence, or list of sequences of tokens, produce a collection of directed
```
Comment (jc-healy) [inline vectorizers/timed_token_cooccurrence_vectorizer.py:240]: We should probably include a few details about the (token, time) pair, that this class is designed to tackle, in the docstring

**Review by jc-healy — COMMENTED** [review]
Wow!  This is a really epic PR Colin.  The code looks solid and I love the refactoring of cooccurrence vectorizers to inherit from a base class.  
My only real suggestion is that we need to flesh out the docstrings of the various cooccurrence classes to make it clear when they should be used.  My docstring sense detects some exhaustion and cut and paste.  That is fully justified given the scope of this refactor.

**Review by jc-healy — APPROVED** [review]

**Pull request merged**